
The files generated by the different programs within the LDhat package are all
text files and can be manipulated by standard software. However, scripts for
summarising the results using the R software (see http://www.r-project.org)
are also available. To obtain these, start an R session in the directory where
the programs have been run (or change directory to get there) and type

source("https://raw.githubusercontent.com/auton1/LDhat/master/coalescent.r")

Additional details can be specified with the options
summary<-summarise.interval(rates.file = "Mbel_10_samples_contig_tarseq_0.vcf.beagle.ldhat.bpen2.rates.txt",
                            locs.file="Mbel_10_samples_contig_tarseq_0.vcf.beagle.ldhat.locs")


data_LD <- read.table("PopLDdecay.Mbel_SNP_biallelic_concat.vcf.gz_ph_imp_10kb.stat.bin_duplicate.txt",header = F, sep = "\t")
head(data_LD)
set_plot_dimensions(6, 4)
dev.new(width=5, height=5, unit="cm") # cm # https://www.tutorialspoint.com/how-to-create-a-plot-in-r-with-a-different-plot-window-size-using-plot-function
# https://www.geeksforgeeks.org/create-plot-window-of-particular-size-in-r/
plot(data_LD[,1]/1000,data_LD[,2], bty="L",lwd=1,type = "l",col="blue",main = "LD decay of M_bel", xlab = "Distance (Kb)", ylab = expression(r^2))
dev.off()
axis(2, at=1:5)
par(new=T)
library(gtools)
mixedsort(list.files())
mixedsort

data_400kb <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_400kb_windows_sorted_new.txt", header = T,sep = "\t")
head(data_400kb)
dev.new(width=8, height=4, unit="in")
par(mar=c(4,6,1.5,2))
plot(data_400kb$GEN_MID/1000000,data_400kb$rho_per_kb, type = "l", xlab = "Genome position (Mb)", ylab = "Recombination rate\nrho/kb")



# plot for contig_0
rho_tarseq_0 <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen=5_contig_0.txt", header = F, sep = "\t")
head(rho_tarseq_0)
library(ggplot2)
plot_0 <- ggplot(rho_tarseq_0,aes(V2,V3))+
  geom_line(size=0.2)+
  scale_x_continuous(expand=c(0,0))+
  scale_y_continuous(expand=c(0,0))+
  labs(x="Contig (kb)",y="rho/bp",title = "Recombination rate across region")+ # Remove space between plotted data and the axes https://stackoverflow.com/questions/22945651/remove-space-between-plotted-data-and-the-axes
  theme_classic()+
  theme(plot.title = element_text(hjust = 0.5, vjust = 0.1, size = 11),
        axis.text = element_text(colour = "black",size = 10),
        axis.line = element_line(size = 0.3))

last_plot() # return the last plot
ggsave("contig_0_recombination rate.pdf",width = 5,height = 4) # save figure in the working dir

plot_1 <- ggplot(rho_tarseq_0,aes(V2))+
  geom_histogram(binwidth =100)


####
Mbel_10_samples_contig_tarseq_0.vcf.beagle.ldhat.bpen5.res.txt
num 
file_list <- as.data.frame((list.files(pattern = "*.res")))
head(file_list,1)
print(file_list)
write.table(file_list,file="file_list.txt",sep = "\t",quote = F,row.names = F,col.names = F)

data_contig_0_5 <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_tarseq_${}_tab.txt", header = T, sep = "\t")

#### split a combined file
Mbel_whole_genome_LDhat_stat_out_bpen_2_new_sorted_tab.txt
data_contig_combined_1 <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_new_sorted_tab.txt", nrows = 17, header = T, sep = "\t")
data_contig_combined_all <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_new_sorted_tab.txt", header = T, sep = "\t")
#  create a subset data
subset_all_first_5 <- subset(data_contig_combined_all, select=c(1:3),data_contig_combined_2$Contig==c(head(unique(data_contig_combined_2$Contig)))) # the first 10
head(subset_all_first_5)
tail(subset_all_first_5)
subset_all <- subset(data_contig_combined_all, select=c(1:3))
head(subset_all)

create_dir('Mbel_contig_mean_rho')   # create directory for this
contig_name_all_split <- split(subset_all, f=subset_all$Contig)
head(contig_name_all_split)
contig_name_all <- names(contig_name_all_split) # create separate files

setwd("G:/NGS_1/Sweden/recombination/summary/Mbel_contig_mean_rho/")
list.files()
for (contig in contig_name_all){
  saveName = paste0(contig, '.Rdata')
  saveRDS(contig_name_all_split[[contig]],file = saveName) # save file to .RData file
}

readRDS("tarseq_1236.Rdata")

#### read all files
data_list_all = list.files(pattern = "*.Rdata")
head(data_list_all)
tail(data_list_all)
str(data_list_all)

install.packages("gtools")
library(gtools)
sorted_contig <- mixedsort(data_list_all) # sort strings with numeric and alphabetic letters
head(sorted_contig, 15)
tail(sorted_contig)
str(sorted_contig)
sorted_contig_per_line <- as.data.frame(sorted_contig)
head(sorted_contig_per_line)
tail(sorted_contig_per_line)

####
data_contig_all= lapply(sorted_contig_per_line$sorted_contig, readRDS) # read all files; not read.table but readRDS for .Rdata files
class(data_contig_all)
str(data_contig_all) ## total number is 1097
tail(data_contig_all)
head(data_contig_all[[1095]])
head(data_contig_all[[1235]])
data_contig_all[order(as.character(data_contig_all))]

data_contig_all[[1]] # show the first element of the list
head(data_contig_all[[1]])
tail(data_contig_all[[1]])
head(data_contig_all[[1]][,1])
tail(data_contig_all[[2]][,1])
tail(data_contig_all[[2]][,2])
plot(data_contig_all[[1]][,2],data_contig_all[[1]][,3],type = "l")
plot(data_contig_all[[2]][,2],data_contig_all[[2]][,3],type = "l")

#### read multiple file
setwd("G:/NGS_1/Sweden/recombination/summary/Mbel_contig_mean_rho/")
install.packages("gtools")
library(gtools)
temp_contig <- list.files(pattern = "*.Rdata$")
head(temp_contig)
temp <- mixedsort(temp_contig) # sort strings with numeric and alphabetic letters
temp[1]
head(temp)

for (i in 1:length(temp)) assign(temp[i],readRDS(temp[i]))
temp[11]
temp[1]
temp[2]
temp[3]
readRDS(temp[3])
#### second call 'for loop'
library("stringr")

  for (i in c(1:3)){
  data_i <- readRDS(temp[i])
  win_limits <- seq(from = 1, to = max(data_i[,2])+1, by = 1)
    window <- win_func(data_i$Loci)
    # bin_end <- 
  data_i_window <- cbind(data_i,window)
  win_means <- aggregate(data_i$Mean_rho, list(window), FUN = "mean") # https://r-coder.com/aggregate-r/
  contig_i <- as.vector(rep(unique(data_i$Contig), length(unique(window)))) # contig column: https://r-coder.com/aggregate-r/
  contig_win_means <- cbind(contig_i, win_means)
  out <- print(contig_win_means)
  write.table(out, file = paste0(temp[i], "_1K_rho.txt"), sep="\t", row.names = F, quote = F) # print(head(win_means))
  plot(contig_win_means$Group.1, contig_win_means$x, type = "l", xlab = paste0(str_sub(temp[i], start = 1, end = 9), "_position"), ylab = "rho/kb") # output different length of string in the x-axis; https://www.gastonsanchez.com/r4strings/stringr-basics.html
}


print(win_means[3])
str(win_means)
str(data_i)
str(window)
str(data_i_window)

#### test 3 contig 'for loop' & combine files ## tarseq_0.Rdata_1K_rho.txt

getwd()
setwd("G:/NGS_1/Sweden/recombination/summary/Mbel_contig_mean_rho/")
result <- mixedsort(list.files(pattern = "*Rdata_1K_rho.txt"))
head(result)
read_all_results <- lapply(result, function(x){read.table(file=x, header = T,sep = "\t")})
str(read_all_results)
combined_read_all_results <- do.call("rbind", lapply(read_all_results, as.data.frame))
str(combined_read_all_results)
head(combined_read_all_results)
tail(combined_read_all_results)
write.table(combined_read_all_results, file = "combined_read_all_results_contig_0_1_2_1kb_mean_rho.txt",quote = F,row.names = F)
mean(combined_read_all_results$x) # the average rho = 1.596334
summary(combined_read_all_results$x)

#### contig length cumulation file
setwd("G:/NGS_1/Sweden/recombination/summary/")
cum_contig <- read.table("Mbel_contig_cum_length_tab_modified_2.txt", header = T,sep = "\t")
head(cum_contig)
cum_contig$pos[2]
cum_contig$pos[3]
#### example # https://stackoverflow.com/questions/21818427/how-to-add-a-cumulative-column-to-an-r-dataframe-using-dplyr
setwd("G:/NGS_1/Sweden/recombination/summary/Mbel_contig_mean_rho/")
library('dplyr')
df <- data.frame(id=rep(1:3, each=5),
                 hour=rep(1:5,3),
                 value=sample(1:15))
df
head(df)
str(df)
mutate(group_by(df,id), csum=cumsum(value))
mutate(group_by(df,id), csum=cumsum(value)+10)


for (i in cum_contig$pos){
  mutate(group_by(combined_read_all_results,contig_i), contig_cum=cumsum(combined_read_all_results$Group.1+(cum_contig$pos[20]/1000)))
}

str(combined_read_all_results)
str(combined_read_all_results$contig_i)
#### count the number of a vector in a column
str_count(combined_read_all_results$contig_i, "tarseq_2")
length(which(combined_read_all_results$contig_i=="tarseq_1")) # https://stackoverflow.com/questions/1923273/counting-the-number-of-elements-with-the-values-of-x-in-a-vector; https://www.programmingr.com/count-occurrences-in-column/
table(combined_read_all_results$contig_i)


tail(combined_read_all_results$contig_i)
str(combined_read_all_results$Group.1)
head(combined_read_all_results)
tail(combined_read_all_results)

#$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$
data <- read.table("combined_read_all_results_contig_0_1_2_1kb_mean_rho.txt", sep = "\t", header = T)


mutate(group_by(combined_read_all_results,contig_i), contig_cum=cumsum(combined_read_all_results$Group.1)+10)

# mutate(group_by(combined_read_all_results,contig_i), contig_cum=cumsum(combined_read_all_results$Group.1+(contig_length[i]/1000)))





for (i in c(1:3)){
  data_i <- readRDS(temp[i])
  win_limits <- seq(from = 1, to = max(data_i[,2])+1, by = 1)
  window <- win_func(data_i$Loci)
  # bin_end <- 
  data_i_window <- cbind(data_i,window)
  win_means <- aggregate(data_i$Mean_rho, list(window), FUN = "mean")
  out <- print(win_means)
}


  
#### Mats ####
setwd("G:/NGS_1/Sweden/recombination/summary/")
library('dplyr')
data_contig_0 <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_tarseq_0_tab.txt", header = T, sep = "\t")
head(data_contig_0)
class(data_contig_0)
str(data_contig_0)
max(data_contig_0[,2]) # show the biggest number of Loci in 2nd column
win_limits <- seq(from = 1, to = max(data_contig_0[,2])+1, by = 1) # set the range of windows
head(win_limits)
tail(win_limits)
win_func <- Vectorize(function(x){min(which(win_limits > x ))}) # which shows the positions of x less than win_limits; min find the minimum win that > x

#win.function <- function(x)
  
data_contig_0$window <- win_func(data_contig_0$Loci)

head(data_contig_0,+20)
win_means <- aggregate(Mean_rho~window, data = data_contig_0, FUN = "mean")
count <-  aggregate(Mean_rho~window, data = data_contig_0, FUN = "sum")
head(win_means)
require(dplyr)
head(count(data_contig_0,"window"))
win_means$Count <- count(data_contig_0,"window")
head(win_means)


#### contig length cumulation file
cum_contig <- read.table("Mbel_contig_cum_length_tab_modified_2.txt", header = T,sep = "\t")
head(cum_contig)


subset_0 <- subset(data_contig_combined_2, select=c(1:3),data_contig_combined_2$Contig=="tarseq_0")
tail(subset_0)
subset_1 <- subset(data_contig_combined_2, select=c(1:3),data_contig_combined_2$Contig=="tarseq_1")
tail(subset_1)
combined_data_0_1 <- rbind(tail(subset_0),tail(subset_1),make.row.names=F) # combine rows without showing line number
head(combined_data_0_1)
data_split <- split(combined_data_0_1, f=combined_data_0_1$Contig)

write.table(data_split,file="data_split_contig_0_1.txt",quote = F,row.names = F,sep = "\t")
# save(data_split,file = "data_split.RData")
readRDS("data_split.RData")
rm("data_split.RData")

contig_name <- names(data_split)
for (contig in contig_name){
  saveName = paste0(contig, '.Rdata')
  saveRDS(data_split[[contig]],file = saveName) # save file to .RData file
}

readRDS("tarseq_1.Rdata") # read this specific file
#### read multiple file at once
data_list_0 <- read.csv("G:/NGS_1/Sweden/recombination/summary/file_list.txt",header = F)
str(data_list_0)
data_contig_0 <- lapply(data_list_0[1,], read.table,header = T) # read one file each time
setwd("G:/NGS_1/Sweden/recombination/summary/")
data_list = list.files(pattern = "*.res")
data_contig= lapply(data_list, read.table)
str(data_list)
str(data_contig)


#### use R commands
dir.create("test_R") # http://theautomatic.net/2018/07/11/manipulate-files-r/
file.create("G:/NGS_1/Sweden/recombination/summary/test_R/test_R.txt")
sapply(paste0("file",1:10,".txt"), file.create)
file.create("G:/NGS_1/Sweden/recombination/summary/test_R/test_R.txt")
fileSnapshot() # give details about the files in the directory
file.info("data_split.RData")
#### copy a file
file.copy("tarseq_0.Rdata","G:/NGS_1/Sweden/recombination/summary/test_R")

#### delete a file or directory
file.remove("data_split.RData")
unlink("data_split.RData")
unlink("directory",recursive = T)
sapply(paste0("file",1:10,".txt"),unlink) # delete many files
file.exists("data_split.RData") # check if a file exists
basename("tarseq_0.Rdata")
dirname("tarseq_0.Rdata")

#### physically open a file
shell.exec("file_list.txt")
file.show("file_list.txt")
# open a file selection window
file.choose()
# move a file
install.packages("filesstrings")
library(filesstrings)
file.move("C:/path/to/file/some_file.txt", "C:/some/other/path")

# without the 1st line
ls() # list all variables
rm(list=ls()) # remove all variables in memory

data_contig_combined <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_tarseq_0_tab.txt", header = T, sep = "\t")
head(data_contig_combined)
split <- split(data_contig_combined, f=data_contig_combined)
tail(split$tarseq_1)

#### Mats
setwd("G:/NGS_1/Sweden/recombination/summary/")
data_contig_0 <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_tarseq_0_tab.txt", header = T, sep = "\t")
head(data_contig_0)
str(data_contig_0)
max(data_contig_0[,2]) # show the biggest number of Loci in 2nd column
win_limits <- seq(from = 1, to = max(data_contig_0[,2])+1, by = 1) # set the range of windows
head(win_limits)
tail(win_limits)
win_func <- Vectorize(function(x){min(which(win_limits > x ))}) # which shows the positions of x less than win_limits; min
win.function <- function(x)
# data_contig_0$window <- win_func(data_contig_0$Loci)
window <- win_func(data_contig_0$Loci)
win_means <- aggregate(data_contig_0$Loci, list(window), FUN = "mean")
head(win_means)
str(win_means)
str(window)
#### add contig column
contig_0 <- as.vector(rep(unique(data_contig_0$Contig), length(unique(window))))
str(contig_0)
head(contig_0)

contig_win_means <- cbind(contig_0,win_means)
head(contig_win_means)
tail(contig_win_means)

#### re-check

########################################
head(data_contig_0)
str(data_contig_0)
dim(data_contig_0)
max(data_contig_0[,2])
win_limits <- seq(from = 0, to = max(data_contig_0[,2]), by = 1)
head(win_limits)
min(which(win_limits > data_contig_0[1,2] )) # Loci[1,2]=0.074;vector of 1st row and 2nd column;
# which shows the indices of win_limits > a specific loci; 
# min shows the minimize loci among them

win_limits <- seq(from = 1, to = max(data_contig_0[,2])+1, by = 1)
min(which(win_limits > data_contig_0[1,2] ))
head(win_limits)

min(which(win_limits > data_contig_0[500,2] )) # win-226
data_contig_0[500,2] # 225.082

min(which(win_limits > data_contig_0[1:3,2] ))
win_func <- function(x){min(which(win_limits > data_contig_0[x,2] ))} 
# create a function and call the function using "win_func(x)"
win_func(x) # call function
win_func(3)
win_func(10)
win_func(30)
win_func(500)

win_func <- Vectorize(function(x){min(which(win_limits > data_contig_0[x,2] ))}) # x is the vectors in 2nd column 
win_func(c(3,5,30,500)) # check after being vectorized

#data_contig_0$window <- win_func(data_contig_0$Loci) # issue ! -> test x not "loci" variable; assign windows to the data variable
#head(data_contig_0)
#win_func(data_contig_0$Loci[1:5])
#class(data_contig_0$Loci)

win_func <- Vectorize(function(x){min(which(win_limits > x ))}) # x is the vectors in 2nd column
data_contig_0$window <- win_func(data_contig_0$Loci) # x=data_contig_0$Loci
head(data_contig_0)
head(data_contig_0, 20)
tail(data_contig_0, 20)
?aggregate
aggregate()
win_means <- aggregate(Mean_rho~window, data = data_contig_0, FUN = "mean")
# alternative
# win_means <- aggregate(data_contig_0$Mean_rho, FUN=mean, list(data_contig_0$window))
head(win_means)
data_contig_0[data_contig_0$window == 1,] # show the whole row
data_contig_0$Mean_rho[data_contig_0$window == 1] # window for mean_rho of 1
mean(data_contig_0$Mean_rho[data_contig_0$window == 1]) # manually check the mean of window of 1
plot(x = win_means$window-0.5, y = win_means$Mean_rho)
plot(x = win_means$window, y = win_means$Mean_rho, type="l")
?plot

#### combined data
window_1_kb_mbel <- read.table("Mbel_whole_genome_LDhat_stat_out_bpen_2_new_sorted_tab.txt", header = T,sep = "\t")
head(window_1_kb_mbel)
tail(window_1_kb_mbel)


#### perl results 20 kb windows
window_20_kb_mbel <- read.table("G:/NGS_1/Sweden/recombination/summary/Mbel_whole_genome_LDhat_stat_out_bpen_2_20kb_windows_sorted_new.txt", header = T,sep = "\t")
head(window_20_kb_mbel)
plot(window_20_kb_mbel$GEN_MID, window_20_kb_mbel$rho_per_kb, type="l")

#### write own function() with arguement and call it
new.function <- function(a){
  for (i in 1:a) {
    b<- i^2
    print(b)
  }
}

new.function(5)

#### write own function() without arg and call it
new.function <- function(){
  for (i in 1:6){
    b <- i^3
    print(b)
  }
}

new.function()

new.function <- function(a,b,c){
  result <- a*b+c
  print(result)
}

new.function(2,3,4)

### CpGoe
setwd("G:/NGS_1/Sweden/CpG")
ls()
rm(list=ls()) # remove all existed variable
list.files()
data_cpg <- read.table("C_sec_GC_content_CpG_new.txt", sep = "\t", header = T)
head(data_cpg)
tail(data_cpg)  
library(ggplot2)
#### 1
plot_cpg <- ggplot(data_cpg,aes(GC_content,CpGoe))+
  geom_point(size=0.2,stroke=0,alpha = 0.5)+
  theme_classic()+
  theme(plot.margin = margin(8,8,8,8,"pt")) # alpha = 1/5

ggsave(plot_cpg,file = "C_sec_data_cpg_1.pdf")
#### 2
plot_cpg_pos <- ggplot(data_cpg,aes(win_pos,CpGoe))+
  geom_line()+
  theme_classic()+
  theme(plot.margin = margin(8,8,8,8,"pt"))

ggsave(plot_cpg_pos,file = "C_sec_data_cpg_pos.pdf")
#### 3 - choose a better value for bins = ?
plot_cpg_hist <- ggplot(data_cpg,aes(CpGoe))+
  geom_histogram(binwidth = 0.005,color="red",fill="white",size=0.05)+
  theme_classic()+
  labs(title = "Histogram of CpGo/e",x="CpGo/e",y="Count")+
  theme(plot.title = element_text(hjust = 0.5,vjust = 1,size = 11),
        axis.text = element_text(color = "black",size = 10), 
        axis.line = element_line(size = 0.3),
        plot.margin = margin(8,8,8,8,"pt"))

ggsave(plot_cpg_hist,file = "C_sec_data_cpg_hist-.pdf")

